pygbm/gradient_boosting.py

```diff
@@ -244,6 +242,10 @@ def fit(self, X, y):
                 _update_raw_predictions(leaves_data, raw_predictions[:, k])
                 toc_pred = time()
                 acc_prediction_time += toc_pred - tic_pred
+                # make sure we collect memory
+                del grower
+                import gc
+                gc.collect()
```

Comment:
I would rather not call `gc.collect()` explicitly in a library. It can induce significant overhead if the python process has large non-collectable objects collections in memory (potentially outside of the library code).

If we want to improve the memory further here we could instead try to find a way to reuse large internal datastructures of the grower object although this would probably make the code less readable.

Reply:
What causes the stress is most probably the allocation of the splitter. I think it would be possible to allocate it only once in `fit()`, and pass it as an attribute to each of the `n_iter` growers.
Alternatively we could allocate only one grower, and instead of using `grower = TreeGrower(...)`, we could do `grower.reset(...)`

Reply:
Yes I think it would make sense from a memory usage point of view.

If someone decides to work on this, it would be great to run the higgs boson benchmark script (with --no-lightgbm) under `memory_profiler` to check the impact of the refactoring.

Reply:
Yes, gc.collect isn't great, reusing would be much nicer. It is hard for me to benchmarks, since it takes a significant amount of time to run. The difference is noticeable, I can type now, instead of the computer being seriously slow such that typing is even hard. I saw memory peaks going to 28+GB, and now it is more stable at 21 GB

Reply:
What is the dataset size? On the Higgs boson benchmark, memory usage stays around 6GB on my laptop.

Reply:
Or maybe you grow very large trees?

Reply:
it's the iris dataset, but now repeated so it has 10^9 rows, 4 features/column, it is also using some unreleased code (vaex-ml) that does some extra dirty tricks to make pygbm play nice with vaex.

Reply:
I got confused by that as well. I thought Python only does cyclic GC's routinely, and the rest is just reference counts, if 0, GC it. But without gc.collect I see mem usage peaks, and everything just behaves very nondeterministically. So I assumed for a moment that Python just collect every now and then, but it should not.

In any case, reusing the memory seems like the best solution, should I remove that commit?

Reply:
+1